content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md

```diff
@@ -152,6 +152,17 @@ Pick a control plane node that you wish to upgrade first. It must have the `/etc
    [upgrade/kubelet] Now that your control plane is upgraded, please proceed with upgrading your kubelets if you haven't already done so.
    ```
 
+   {{< note >}}
+   Before v1.27 (including v1.27), kubeadm will upgrade the addons (including CoreDNS and kube-proxy) immediately during 
+   `kubeadm upgrade apply`, regardless of whether there are other control plane instances that have not been upgraded, which 
+   may cause compatibility problems. Since v1.28, kubeadm will always checks whether all the control plane instances have 
+   been upgraded before starting to upgrade the addons. You **MUST** perform control plane nodes upgrade sequentially, and 
```

Comment:
@neolit123 Because the check is only performed during `kubeadm upgrade apply/node`, when the control plane nodes are upgrading in parallel, there may be a race that all the kubeadm upgrade processes assume that there are unupgraded control plane instances, this will cause the addons to miss performing the upgrade. Complete serial execution is not required, but in order to ensure that the addons can be successfully upgraded during the last cp node upgrade, users should ensure that all the other cp nodes have been upgraded before doing the last cp node upgrade.  In the future, we can separate the addons upgrade into a separate phase and call it after all the cp nodes have been upgraded.

Reply:
If parallel cp upgrade is necessary, the user needs to make additional checks to see if the addons have been successfully upgraded or always to call `kubeadm upgrade node` on the last cp node again to ensure that the addons can be upgraded successfully. This should be safe because the upgrade operation is idempotent.

Reply:
thanks for the clarification.